src/core_scan/Core_scan.ml

```diff
@@ -702,57 +706,60 @@ let iter_targets_and_get_matches_and_exn_to_errors (config : Core_scan_config.t)
-         AST_generic.version lang file)
-  in
-  {
-    Xtarget.file;
-    xlang;
-    lazy_content = lazy (UFile.read_file file);
-    lazy_ast_and_errors;
-  }
-
-let lockfile_target_of_input_to_core
-    ({ path; lockfile_kind; manifest_target } : In.lockfile_target) =
+let lockfile_target_of_location
+    ({ source; file; kind = lockfile_kind; manifest } :
+      Target_location.lockfile) =
   let manifest_target =
-    manifest_target
-    |> Option.map @@ fun In.{ path; manifest_kind } ->
-       let path = Fpath.v path in
+    manifest
+    |> Option.map
+       @@ fun ({ source; file; kind = manifest_kind } :
+                Target_location.manifest) : Lockfile_target.manifest_target ->
        {
-         Lockfile_target.manifest = path;
-         lazy_manifest_content = lazy (UFile.read_file path);
-         lazy_manifest_ast_and_errors =
-           lazy (Parse_lockfile.parse_manifest manifest_kind path);
-         manifest_kind = Manifest_kind.of_lockfile_kind lockfile_kind;
+         source;
+         file;
+         lazy_content = lazy (UFile.read_file file);
+         lazy_ast_and_errors =
+           lazy (Parse_lockfile.parse_manifest manifest_kind file);
+         kind = Manifest_kind.of_lockfile_kind lockfile_kind;
        }
   in
-  let path = Fpath.v path in
   {
-    Lockfile_target.lockfile = path;
-    lockfile_kind;
-    lazy_lockfile_content = lazy (UFile.read_file path);
-    lazy_lockfile_ast_and_errors =
-      lazy (Parse_lockfile.parse_lockfile lockfile_kind manifest_target path);
-    manifest_target;
+    Lockfile_target.source;
+    file;
+    kind = lockfile_kind;
+    lazy_content = lazy (UFile.read_file file);
+    lazy_ast_and_errors =
+      lazy (Parse_lockfile.parse_lockfile lockfile_kind manifest_target file);
+    manifest = manifest_target;
   }
 
+let manifest_target_location_of_input_to_code
+    ({ path; manifest_kind = kind } : In.manifest_target) :
+    Target_location.manifest =
+  Target_location.manifest_of_source kind (File (Fpath.v path))
+
+let lockfile_target_location_of_input_to_code
+    ({ path; lockfile_kind = kind; manifest_target } : In.lockfile_target) :
+    Target_location.lockfile =
+  let manifest =
+    Option.map manifest_target_location_of_input_to_code manifest_target
+  in
+  Target_location.lockfile_of_source ?manifest kind (File (Fpath.v path))
+
+let code_target_location_of_input_to_code
+    ({ path; analyzer; products; lockfile_target } : In.code_target) :
+    Target_location.code =
+  let lockfile =
+    Option.map lockfile_target_location_of_input_to_code lockfile_target
+  in
+  Target_location.code_of_source ?lockfile analyzer products
+    (File (Fpath.v path))
+
+let target_location_of_input_to_code (input : In.target) : Target_location.t =
+  match input with
+  | `CodeTarget x -> Code (code_target_location_of_input_to_code x)
+  | `LockfileTarget x -> Lockfile (lockfile_target_location_of_input_to_code x)
```

Comment:
I don't about this type that uses polymorphic variants but I suspect classic variants are better. I prefer them anyway. The reasons are:
* simplicity by consistency (why use two tools when we could use just one)
* less confusing error messages, especially when type annotations are lacking
Reasons for using polymorphic variants include:
* extensibility, subtyping - which is occasionally useful

Reply:
Yep these are from `Input_to_core.atd`. This should also probably be turned into a record with shared fields and then a field with a sum.

Reply:
I will defer to whatever Matthew wants to do here, but since this would require an additional interfaces change I think it best to do in a separate PR, especially as this PR does not introduce these types. Resolving here.